### PR TITLE
SMCCC TRNG driver (and embed in vexpress-juno)

### PR DESCRIPTION
**Pull request by etienne-lms**
This implements a RNG driver based on Arm SMCCC TRNG interface.
Since this TF-A Juno platform exposes this feature (since v2.5), default enable the OP-TEE driver for that platform.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
It works with this diff applied:
``` diff
diff --git a/core/drivers/smccc_trng.c b/core/drivers/smccc_trng.c
index caf9caad0d7b..77808b7d31fb 100644
--- a/core/drivers/smccc_trng.c
+++ b/core/drivers/smccc_trng.c
@@ -36,6 +36,11 @@
 #define TRNG_MAJOR_SHIFT	16
 #define TRNG_MINOR_MASK		GENMASK_32(15, 0)
 #define TRNG_MINOR_SHIFT	0
+#define TRNG_MAKE_VERSION(major, minor)	\
+	((SHIFT_U32(major, TRNG_MAJOR_SHIFT) & TRNG_MAJOR_MASK) | \
+	 (SHIFT_U32(minor, TRNG_MINOR_SHIFT) & TRNG_MINOR_MASK))
+
+#define TRNG_VERSION_1_0	TRNG_MAKE_VERSION(1, 0)
 
 #define TRNG_MAX_RND_64		(192 / 8)
 #define TRNG_MAX_RND_32		(96 / 8)
@@ -65,25 +70,26 @@ static bool smccc_trng_is_supported(void)
 
 	inited = true;
 
-	/* Feature discovery require Arm SMCCC version >= v1.1 */
+	/*
+	 * TRNG ABI requires caller to check that Arm SMCCC version is
+	 * larger or equal to v1.1
+	 */
 	args.a0 = ARM_SMCCC_VERSION;
 	thread_smccc(&args);
-	if (args.a0 | BIT32(31))
+	if (args.a0 & BIT32(31))
 		return false;
 	if (args.a0 < SMCCC_V_1_1)
 		return false;
 
-	/* Test TRNG function IDs support */
-	args.a0 = ARM_SMCCC_ARCH_FEATURES;
-	args.a1 = ARM_SMCCC_TRNG_VERSION;
+	/*
+	 * Check TRNG version, if success we're guarenteed to have at least
+	 * the ARM_SMCCC_TRNG_FEATURES fid.
+	 */
+	args.a0 = ARM_SMCCC_TRNG_VERSION;
 	thread_smccc(&args);
-	if (args.a0 != ARM_SMCCC_RET_SUCCESS)
+	if (args.a0 & BIT32(31))
 		return false;
-
-	args.a0 = ARM_SMCCC_ARCH_FEATURES;
-	args.a1 = ARM_SMCCC_TRNG_FEATURES;
-	thread_smccc(&args);
-	if (args.a0 != ARM_SMCCC_RET_SUCCESS)
+	if (args.a0 < TRNG_VERSION_1_0)
 		return false;
 
 #ifdef ARM64

File: core/drivers/smccc_trng.c [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+static bool trng_supported;
+/* True if ARM_SMCCC_TRNG_RND_64 is supported, false otherwise */
+static bool trng_smc64b;
+
+struct trng_pdata {
+	uint32_t func_id;	/* SMCCC Function ID to use */
+	size_t sz_max;		/* Max byte size of entropy per invocation */
+};
+
+static const struct trng_pdata trng_pdata[2] = {
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/smccc_trng.c:53]: ok, i'll update.

**Review by etienne-lms — COMMENTED** [review]
Thanks for the feedback and proposal (and fixing my buggy implementation).

**Comment by etienne-lms** [comment]
updated

**Comment by jenswi-linaro** [comment]
Found two more issues fixed with
``` diff
diff --git a/core/drivers/smccc_trng.c b/core/drivers/smccc_trng.c
index 6bfd48318ab1..b94eac3ca5cd 100644
--- a/core/drivers/smccc_trng.c
+++ b/core/drivers/smccc_trng.c
@@ -97,43 +97,31 @@ static bool smccc_trng_is_supported(void)
 	return false;
 }
 
-static size_t __maybe_unused read_64b(uint64_t sample, uint8_t *ptr, size_t len)
+static uint8_t *read_bytes(unsigned long val, size_t byte_count, uint8_t *buf,
+			   size_t *rem)
 {
-	size_t sz = MIN(len, sizeof(sample));
+	size_t n = 0;
 
-	memcpy(ptr, &sample, sz);
+	for (n = 0; n < MIN(byte_count, *rem); n++)
+		buf[n] = val >> (n * 8);
 
-	return len - sz;
-}
-
-static size_t read_32b(uint32_t sample, uint8_t *ptr, size_t len)
-{
-	size_t sz = MIN(len, sizeof(sample));
-
-	memcpy(ptr, &sample, sz);
-
-	return len - sz;
+	*rem -= n;
+	return buf + n;
 }
 
 static void read_samples(struct thread_smc_args *args, uint8_t *ptr, size_t len)
 {
 	size_t rem = len;
+	size_t byte_count = 4;
 
-	/* The below assumes little endian random data */
 #ifdef ARM64
-	if (trng_rnd_fid == ARM_SMCCC_TRNG_RND_64) {
-		rem = read_64b(args->a3, ptr, rem);
-		rem = read_64b(args->a2, ptr, rem);
-		rem = read_64b(args->a1, ptr, rem);
-		assert(!rem);
-		return;
-	}
+	if (trng_rnd_fid == ARM_SMCCC_TRNG_RND_64)
+		byte_count = 8;
 #endif
 
-	rem = read_32b(args->a3, ptr, rem);
-	rem = read_32b(args->a2, ptr, rem);
-	rem = read_32b(args->a1, ptr, rem);
-	assert(!rem);
+	ptr = read_bytes(args->a3, byte_count, ptr, &rem);
+	ptr = read_bytes(args->a2, byte_count, ptr, &rem);
+	ptr = read_bytes(args->a1, byte_count, ptr, &rem);
 }
 
 static TEE_Result __maybe_unused smccc_trng_read(void *buf, size_t len)
@@ -155,7 +143,7 @@ static TEE_Result __maybe_unused smccc_trng_read(void *buf, size_t len)
 		size_t burst = MIN(rem, max_burst);
 
 		args.a0 = trng_rnd_fid;
-		args.a1 = burst;
+		args.a1 = burst * 8;
 
 		thread_smccc(&args);
 
```

**Review by jenswi-linaro — COMMENTED** [review]
`Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (Juno)`
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
Squashed fixup commits and applied review tags.
Thanks @jenswi-linaro for the support.

**Pull request merged**